File: Searching/FibonacciSearch/fibonacciSearch.py
def fibonacci_search(arr, target):
    n = len(arr)
    fib_m_minus_2 = 0
    fib_m_minus_1 = 1
    fib_m = fib_m_minus_1 + fib_m_minus_2
    
    while fib_m < n:
        fib_m_minus_2 = fib_m_minus_1
        fib_m_minus_1 = fib_m
        fib_m = fib_m_minus_1 + fib_m_minus_2
    
    offset = -1
    
    while fib_m > 1:
        i = min(offset + fib_m_minus_2, n - 1)
        if arr[i] < target:
            fib_m = fib_m_minus_1
            fib_m_minus_1 = fib_m_minus_2
            fib_m_minus_2 = fib_m - fib_m_minus_1
            offset = i
        elif arr[i] > target:
            fib_m = fib_m_minus_2
            fib_m_minus_1 = fib_m_minus_1 - fib_m_minus_2
            fib_m_minus_2 = fib_m - fib_m_minus_1
        else:
            return i
    
    # Check the last element
    if fib_m_minus_1 == 1 and arr[offset + 1] == target:
        return offset + 1
    
    return -1

if __name__ == "__main__":
    arr = [1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21]
    target = int(input("Enter the target element to search: "))

    result = fibonacci_search(arr, target)

    if result != -1:
        print(f"Element found at index {result}")
    else:
        print("Element not found in the array.")

File: Searching/FibonacciSearch/fibonacciSearch.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int fibonacciSearch(const std::vector<int>& arr, int target) {
    int n = arr.size();

    int fibMMinus2 = 0;
    int fibMMinus1 = 1;
    int fibM = fibMMinus1 + fibMMinus2;

    while (fibM < n) {
        fibMMinus2 = fibMMinus1;
        fibMMinus1 = fibM;
        fibM = fibMMinus1 + fibMMinus2;
    }

    int offset = -1;

    while (fibM > 1) {
        int i = std::min(offset + fibMMinus2, n - 1);

        if (arr[i] < target) {
            fibM = fibMMinus1;
            fibMMinus1 = fibMMinus2;
            fibMMinus2 = fibM - fibMMinus1;
            offset = i;
        } else if (arr[i] > target) {
            fibM = fibMMinus2;
            fibMMinus1 = fibMMinus1 - fibMMinus2;
            fibMMinus2 = fibM - fibMMinus1;
        } else {
            return i; // Found the target
        }
    }

    // Check the last element
    if (fibMMinus1 == 1 && arr[offset + 1] == target) {
        return offset + 1;
    }

    return -1; // Target not found
}

int main() {
    std::vector<int> arr = {1, 3, 5, 7, 9, 10 , 11, 13, 15, 17, 19, 21};
    int target;

    std::cout << "Enter the target element to search: ";
    std::cin >> target;

    int result = fibonacciSearch(arr, target);

    if (result != -1) {
        std::cout << "Element found at index " << result << std::endl;
    } else {
        std::cout << "Element not found in the array." << std::endl;
    }

    return 0;
}
